target/linux/ipq40xx/image/generic.mk

```diff
@@ -723,6 +723,21 @@ define Device/linksys_mr8300
 endef
 TARGET_DEVICES += linksys_mr8300
 
+define Device/linksys_whw03
+	$(call Device/FitzImage)
+	DEVICE_VENDOR := Linksys
+	DEVICE_MODEL := WHW03
+	SOC := qcom-ipq4019
+	KERNEL_SIZE := 8192k
+	IMAGE_SIZE := 131072k
+	IMAGES += factory.bin
+	IMAGE/factory.bin  := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-rootfs | pad-rootfs | linksys-image type=WHW03
+	DEVICE_PACKAGES := ath10k-firmware-qca9888-ct kmod-leds-pca963x kmod-spi-dev kmod-bluetooth \
+		kmod-fs-ext4 e2fsprogs kmod-fs-f2fs mkf2fs losetup \
+		gdisk sgdisk blockdev blkdiscard
```

Comment:
What do you need F2FS support, gdisk, sgdisk, blockdev, blkdiscard for?

Reply:
`e2fsprogs kmod-fs-ext4 losetup` should be enough for eMMC

Reply:
> kmod-fs-ext4 e2fsprogs kmod-fs-f2fs mkf2fs

this is an emmc device. these is the minimum set of packages needed to support https://github.com/openwrt/fstools.
see: https://github.com/openwrt/fstools/blob/08cd7083cac4bddf88459efa0881ee52858e7d0a/libfstools/common.c#L166-L169

> losetup

required to create and mount rootfs_data.

> gdisk sgdisk blockdev blkdiscard

please see the OP.

Reply:
> `e2fsprogs kmod-fs-ext4 losetup` should be enough for eMMC

no, it is not. see the code.

Reply:
mhhh considering the user will have to make manual command and this is a tweak that will make the thing not compatible with stock firmware, i think we can drop these package and instruct the user to install them in the related guide to make the tweak... The user will have to follow some step anyway so one extra step there won't change a thing. DEVICE_PACKAGES usage is to add packages required to have the device fully working, not prepare the device for optional tweak. Hope you understand these reason. (also having all these extra package reduce the space even more for all the user that don't want to expand/repartition the device)

Reply:
You cannot include `gdisk` and `sgdisk` since they are not part of the root repo, also the default device package set should be the minimal required for its HW to function, future disk repartitioning doesnt count here.
When you write/publish that repartitioning/expansion script all of that tooling can be installed via opkg and added as a default then.

Reply:
Regarding F2FS, there is a minimal size limit before fstools defaults to it, I remember it being 100MiB.

Any chance you can post a full bootlog because the only one in the old PR is when it was broken due to UART?

Reply:
@Ansuel no, the packages must be in rootfs, not rootfs_data. installing the packages won't work. also, this is an eMMC device with a GPT that needs to be modified. not giving the user minimum tools to operate the hardware is dangerous: you can have situation in which gdisk is required for recovery or going back to stock. this is less than 0.3 MB in a 3.4 GB device, leaving the tools out make no sense.

Reply:
> not giving the user minimum tools to operate the hardware is dangerous: you can have situation in which gdisk is required for recovery or going back to stock.

example of such situation? If a device has been modified it's user fault and he totally accept the risk of doing so. In a normal case the partition data are not touched I assume.

Aside from that, user can totally use the good firmware selector and generate a custom image online with the require package in rootfs. And as @robimarko said these package are not part of openwrt core and not a chance we move these package from packages to openwrt core.

Honestly we should put priority on having the device merged than pushing for extending the partition of the device. (that can totally come later)

Reply:
> You cannot include `gdisk` and `sgdisk` since they are not part of the root repo, [...] When you write/publish that repartitioning/expansion script all of that tooling can be installed via opkg and added as a default then.

well that is too bad. is this really true? sgdisk cannot be installed via opkg as i said before, it need to be in rootfs. the overlay must be unmounted during the pocedure. if sgdisk cannot be included, users will be stuck with using 3% of the flash.

Reply:
> Regarding F2FS, there is a minimal size limit before fstools defaults to it, I remember it being 100MiB so you are probably right about the f2fs kmod and tooling required here.
> 
> Any chance you can post a full bootlog because the only one in the old PR is when it was broken due to UART?

correct, i tired it (because i hate f2fs) and it doesn't work.

bootlog:
https://gist.github.com/Lanchon/79fdac25753e05c1efbc065ec625be9a

note that that boot has other packages like lvm2 that i am using to test stuff.

Reply:
> well that is too bad. is this really true? sgdisk cannot be installed via opkg as i said before, it need to be in rootfs. the overlay must be unmounted during the pocedure. if sgdisk cannot be included, users will be stuck with using 3% of the flash.

@robimarko please let me know if i need to remove those packages. better a crippled device than no device.

Reply:
@robimarko @Ansuel i removed `gdisk sgdisk blockdev blkdiscard` because they are not part of openwrt core. i didn't know of this requirement. i will solve the partitioning thing somehow down the line.

Reply:
> And as @robimarko said these package are not part of openwrt core and not a chance we move these package from packages to openwrt core.
> 
> Honestly we should put priority on having the device merged than pushing for extending the partition of the device. (that can totally come later)

sorry! i had missed your message